### PR TITLE
Events

### PR DESCRIPTION
**Pull request by SimLV**

### CODE REVIEW

**Comment by Loobinex** [comment]
@SimLV  How do I test this?

**Comment by SimLV** [comment]
I wrote a test tool. I think it should be placed in dktools. I will make a PR there and reference it here

**Comment by SimLV** [comment]
Here is a tool: https://github.com/SimLV/dktestrunner

File: src/frontmenu_ingame_tabs.c [hunk 1]
```diff
@@ -1769,7 +1769,9 @@ void maintain_event_button(struct GuiButton *gbtn)
     if ((dungeon->visible_event_idx != 0) && (evidx == dungeon->visible_event_idx))
     {
         turn_on_event_info_panel_if_necessary(dungeon->visible_event_idx);
-        if (is_game_key_pressed(Gkey_ToggleMessage, &keycode, false))
+        //TODO: that should be not here, Keys should be processed at one place
```
Comment (Loobinex) [inline src/frontmenu_ingame_tabs.c:1772]: Where do you suggest I put this?

Comment (SimLV) [reply src/frontmenu_ingame_tabs.c:1772]: Somewhere near "keys".
This is actually a "button update handler" - it is called from ui drawing code.
In this file should be a callback function called from input-related file or a polling function called from there.

Comment (Loobinex) [reply src/frontmenu_ingame_tabs.c:1772]: Sorry, you might as well be speaking russian here.

File: src/game_legacy.h [hunk 2]
```diff
@@ -339,7 +339,7 @@ long field_1516FB;
     long boulder_reduce_health_wall;
     long boulder_reduce_health_slap;
     long boulder_reduce_health_room;
-    struct GuiMessage messages[GUI_MESSAGES_COUNT];
+    struct GuiMessage messages_[3];
```
Comment (Loobinex) [inline src/game_legacy.h:342]: This one still seems wrong.

File: src/player_computer.c [hunk 3]
```diff
@@ -1512,6 +1513,10 @@ void setup_computer_players2(void)
             skirmish_AI_type = 7;
         }
         setup_a_computer_player(i, skirmish_AI_type);
+        if ((game.flags_font & FFlg_AlexCheat) != 0)
```
Comment (Loobinex) [inline src/player_computer.c:1516]: This seems outdated code too. This is compuchat now, right?

File: tests/campaign.cfg [hunk 4]
```diff
@@ -0,0 +1,93 @@
+[common]
```
Comment (Loobinex) [inline tests/campaign.cfg:1]: I would simply not include this file and entire folder here.  You've got a functioning version at the test-tool, seems like a better spot.

**Comment by Loobinex** [comment]
You're pulling into the console branch, could you pull into the master instead?

File: src/frontmenu_ingame_tabs.c (hunk 1 above)
Comment (SimLV) [reply src/frontmenu_ingame_tabs.c:1772]: I mean we should group all keyboard input towards one layer and one place in codebase.
 This function `maintain_event_button` is only used in GuiButtonInit structure as `maintain callback`. It is a drawing abstraction "lets check how should we draw a specific button". 
But here we implement some input logic instead. 
That is not good.

File: src/game_legacy.h (hunk 2 above)
Comment (SimLV) [reply src/game_legacy.h:342]: Why? I should maintain size of a struct consistent with DK but now it is possible to set GUI_MESSAGES_COUNT to any value.

File: tests/campaign.cfg (hunk 4 above)
Comment (SimLV) [reply tests/campaign.cfg:1]: Yes

File: src/frontmenu_ingame_tabs.c (hunk 1 above)
Comment (Loobinex) [reply src/frontmenu_ingame_tabs.c:1772]: I'm new at all this, so while I understand what you're saying, I don't know how to fix this. I'm open to suggestions.

File: src/game_legacy.h (hunk 2 above)
Comment (Loobinex) [reply src/game_legacy.h:342]: I see what you mean.

File: src/frontmenu_ingame_tabs.c (hunk 1 above)
Comment (SimLV) [reply src/frontmenu_ingame_tabs.c:1772]: Actually we have to make some kind of refactoring. 
Good way to do this is to have 'keyboard layout' abstraction.
I.e. when you are in first person mode you have one layout, when you are in dungeon mode you have another etc.
So it should be two mappings `keyboard shortcut -> binding name` and `binding name -> function`

Then it is good to have a function like`void register_key_event(struct KeyboardLayout *layout, enum BindingName binding, void (*) (void *context), void *context);`
Each "component" should have some "init" function where it should register its events.
So when you have some transition in game state (i.e. player possess a creature or open a menu) Game should just set active layout.
And main loop should just call something like `process_keyboard_event(active_layout, keycode, modifiers);`

File: src/player_computer.c (hunk 3 above)
Comment (SimLV) [reply src/player_computer.c:1516]: Yes!

Comment (Loobinex) [reply src/player_computer.c]: So, will you change it?

File: tests/campaign.cfg (hunk 4 above)
Comment (Loobinex) [reply tests/campaign.cfg:1]: So, could you take these files and folders out of the pull?

File: src/player_computer.c (hunk 3 above)
Comment (SimLV) [reply src/player_computer.c]: fixed

File: tests/campaign.cfg (hunk 4 above)
Comment (SimLV) [reply tests/campaign.cfg]: done

**Comment by Loobinex** [comment]
image

**Comment by Loobinex** [comment]
As a reminder, to build code for autotesting, build with:
`make standard STLOGFLAGS=-DAUTOTESTING=1`

**Pull request merged**